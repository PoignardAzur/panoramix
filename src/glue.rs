//! Bunch of types mostly stolen from Crochet, to use as glue with Druid

#![allow(unused)]

use druid::Data;
use std::collections::HashMap;
use std::sync::Arc;

pub use druid::debug_state::DebugState;

pub struct GlobalEventCx<'a> {
    pub app_data: &'a mut DruidAppData,
}

impl<'a> GlobalEventCx<'a> {
    pub fn new(app_data: &'a mut DruidAppData) -> Self {
        Self { app_data }
    }
}

/// The type we use for app data for Druid integration.
///
/// Currently this is action queues.
///
/// It should probably be a vec of actions, but we can refine
/// later. For button clicks it doesn't matter.
#[derive(Clone, Data, Default)]
pub struct DruidAppData(Arc<HashMap<Id, Action>>);

/// Actions that can be produced by widgets,
#[derive(Clone)]
pub enum Action {
    Clicked,
    FutureResolved,
    TextChanged(String),
}

impl DruidAppData {
    pub(crate) fn queue_action(&mut self, id: Id, action: Action) {
        Arc::make_mut(&mut self.0).insert(id, action);
    }

    pub(crate) fn dequeue_action(&mut self, id: Id) -> Option<Action> {
        if self.0.contains_key(&id) {
            Arc::make_mut(&mut self.0).remove(&id)
        } else {
            None
        }
    }

    /// Report whether the id has a non-empty action queue.
    pub(crate) fn has_action(&self, id: Id) -> bool {
        self.0.contains_key(&id)
    }
}

// ---

use std::sync::atomic::{AtomicUsize, Ordering};

static ID_COUNTER: AtomicUsize = AtomicUsize::new(0);

/// An identifier for an element.
///
/// It's a bit heavy-handed to have this id as well as widget
/// id in Druid; likely the two concepts should be unified. But
#[derive(Clone, Copy, Debug, PartialOrd, PartialEq, Ord, Eq, Hash)]
pub struct Id(usize);

impl Id {
    /// Allocate a new unique id.
    pub fn new() -> Id {
        Id(ID_COUNTER.fetch_add(1, Ordering::Relaxed))
    }
}

// ---

use tracing_subscriber::prelude::*;
use tracing_subscriber::{fmt, EnvFilter};
use tracing_unwrap::ResultExt;

pub fn init_tracing() {
    let fmt_layer = fmt::layer().with_target(true);
    let filter_layer = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("warn"))
        .unwrap_or_log();

    tracing_subscriber::registry()
        .with(filter_layer)
        .with(fmt_layer)
        .init();
}
