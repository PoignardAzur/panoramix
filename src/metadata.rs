use derivative::Derivative;

/// Placeholder type for elements that don't raise events.
///
/// Equivalent to `!`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum NoEvent {}

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash)]
pub struct NoState;

#[derive(Derivative)]
#[derivative(
    Clone(bound = ""),
    Copy(bound = ""),
    Default(bound = ""),
    PartialEq(bound = ""),
    Eq(bound = ""),
    Hash(bound = "")
)]
pub struct Metadata<ComponentEvent, ComponentState> {
    _marker: std::marker::PhantomData<(ComponentEvent, ComponentState)>,
}

impl<ComponentEvent, ComponentState> Metadata<ComponentEvent, ComponentState> {
    pub fn new() -> Self {
        Default::default()
    }
}
