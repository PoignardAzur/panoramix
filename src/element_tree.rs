use crate::glue::{DruidAppData, GlobalEventCx};
use crate::widget_sequence::WidgetSequence;

use druid::{Env, EventCtx};
use std::any::Any;
use std::fmt::Debug;

/// Context type passed to all components when building them.
pub struct CompCtx {
    pub(crate) local_state: Box<dyn Any>,
}

impl CompCtx {
    /// Returns the local state of the current component instance.
    ///
    /// Panics if the generic type doesn't match the component's local state type.
    pub fn use_local_state<T: 'static>(&self) -> &T {
        self.local_state.downcast_ref::<T>().unwrap()
    }

    // TODO - add methods
    // use_lifecycle
    // get_vdom_context
}

/// Context required by [`VirtualDom::reconcile`]
pub struct ReconcileCtx<'a, 'b, 'c, 'd, 'e> {
    pub event_ctx: &'a mut EventCtx<'d, 'e>,
    pub data: &'b mut DruidAppData,
    pub env: &'c Env,
}

/// The trait implemented by all GUI elements.
///
/// Every type you use to explicitly create a GUI in Panoramix ([`Button`](crate::elements::Button), [`TextBox`](crate::elements::TextBox), any user-made component) implements Element. You usually don't need to worry about this trait unless you want to implement your own custom element.
///
/// For helper methods that can be called on all elements, see [`ElementExt`].
///
/// ## Note about template resolution
///
/// This trait is parameterized on two template types: `CpEvent` and `CpState`, which represent the event and local-state type of the parent component an element is built in. They are supposed to flow "inwards" with type inference, starting from the `-> impl Element<MyEvent, MyState>` return type of your function.
///
/// To give a concrete example:
///
/// ```rust
/// # use panoramix::{component, CompCtx, Column, Element, ElementExt};
/// # use panoramix::elements::{ButtonClick, Button, Label};
/// # type BuyItem = ButtonClick;
/// #
/// #[component]
/// fn StoreItem(ctx: &CompCtx, item_name: String) -> impl Element<BuyItem, u32> {
///     let item_count = ctx.use_local_state::<u32>();
///     Column!(
///         Label::new(format!("Item: {} x{}", item_name, item_count)),
///         Button::new("+")
///             .on_click(|item_count, _| {
///                 *item_count += 1;
///             }),
///         Button::new("Buy")
///             .bubble_up::<BuyItem>()
///     )
/// }
/// ```
///
/// In this example, because the return type is declared to be `-> impl Element<BuyItem, u32>`, all elements that are returned (Label, Button, Column) will be transitively inferred to implement `Element<BuyItem, u32>`.
///
/// The flip side of this is that constructing an element and not returning it (eg doing `let x = Button::new("...");` and then not using `x`) will lead to a compile error, because the compiler can't infer what `CpEvent` and `CpState` should be.
///
pub trait Element<CpEvent = NoEvent, CpState = ()>: Debug {
    /// The type of events this element can raise.
    ///
    /// This is the type that [`ElementExt::on`], [`ElementExt::map_event`] and [`ElementExt::bubble_up`] can take. It's different from the `CpEvent` generic parameter, which is the event the parent component emits.
    ///
    /// In the `StoreItem` example, the `Event` type of buttons is `ButtonClick`, and their `CpEvent` parameter is `BuyItem`.
    type Event;

    type AggregateChildrenState: Clone + Default + Debug + PartialEq;
    type BuildOutput: VirtualDom<
        CpEvent,
        CpState,
        Event = Self::Event,
        AggregateChildrenState = Self::AggregateChildrenState,
    >;

    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState);
}

// TODO - Include documentation about what a Virtual DOM is and where the name comes from.
pub trait VirtualDom<CpEvent, CpState>: Debug {
    type AggregateChildrenState: Clone + Default + Debug + PartialEq;
    type TargetWidgetSeq: WidgetSequence;

    type Event;

    // update_value is intended to enable memoize-style HOC
    // where instead of returning a vdom node, it returns
    // something along the lines of struct KeepEverythingAsItWas()
    // Ugh. I'm not explaining this well.
    fn update_value(&mut self, other: Self)
    where
        Self: Sized,
    {
        *self = other;
    }

    fn init_tree(&self) -> Self::TargetWidgetSeq;

    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    );

    // TODO - Rename methods
    #[allow(unused_variables)]
    fn process_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<CpEvent> {
        None
    }

    #[allow(unused_variables)]
    fn process_local_event(
        &self,
        component_state: &mut CpState,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) -> Option<Self::Event> {
        None
    }
}

/// Placeholder type for elements that don't raise events.
///
/// Equivalent to `!`.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum NoEvent {}

// Used in unit tests
#[allow(dead_code)]
pub(crate) fn assign_empty_state_type(_elem: &impl Element<NoEvent, ()>) {}

#[allow(dead_code)]
pub(crate) fn assign_state_type<CpEvent, CpState, Elem: Element<CpEvent, CpState>>(_elem: &Elem) {}

use crate::elements::with_event::{ParentEvent, WithBubbleEvent, WithCallbackEvent, WithMapEvent};

/// Helper methods that can be called on all elements.
pub trait ElementExt<CpEvent, CpState>: Element<CpEvent, CpState> + Sized {
    fn on<EventParam, Cb: Fn(&mut CpState, EventParam)>(
        self,
        callback: Cb,
    ) -> WithCallbackEvent<CpEvent, CpState, EventParam, Self, Cb>
    where
        Self::Event: ParentEvent<EventParam>,
    {
        WithCallbackEvent {
            element: self,
            callback,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
        }
    }

    fn map_event<EventParam, EventReturn, Cb: Fn(&mut CpState, EventParam) -> Option<EventReturn>>(
        self,
        callback: Cb,
    ) -> WithMapEvent<CpEvent, CpState, EventParam, EventReturn, Self, Cb>
    where
        Self::Event: ParentEvent<EventParam>,
        CpEvent: ParentEvent<EventReturn>,
    {
        WithMapEvent {
            element: self,
            callback,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
            _comp_return: Default::default(),
        }
    }

    fn bubble_up<Event>(self) -> WithBubbleEvent<CpEvent, CpState, Event, Self>
    where
        Self::Event: ParentEvent<Event>,
        CpEvent: ParentEvent<Event>,
    {
        WithBubbleEvent {
            element: self,
            _comp_state: Default::default(),
            _comp_event: Default::default(),
            _comp_param: Default::default(),
        }
    }
}

impl<CpEvent, CpState, ET: Element<CpEvent, CpState>> ElementExt<CpEvent, CpState> for ET {}
