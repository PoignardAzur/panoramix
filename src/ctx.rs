use crate::glue::DruidAppData;
use crate::metadata::{Metadata, NoState};
use druid::{Env, EventCtx};
use std::any::{type_name, Any, TypeId};

/// Context type passed to all components when building them.
pub struct CompCtx<'a> {
    // Yeah, we're using a cell, sorry. It's very local, though.
    pub(crate) called_use_metadata: std::cell::Cell<bool>,
    pub(crate) local_state: &'a dyn Any,
}

impl<'a> CompCtx<'a> {
    pub fn use_metadata<ComponentEvent: 'static, ComponentState: 'static>(
        &self,
    ) -> Metadata<ComponentEvent, ComponentState> {
        if self.called_use_metadata.get() {
            panic!("error: 'use_metadata' can only be called once per component")
        }
        self.called_use_metadata.set(true);
        Default::default()
    }

    /// Returns the local state of the current component instance.
    ///
    /// Panics if the generic type doesn't match the component's local state type.
    pub fn get_local_state<ComponentEvent: 'static, ComponentState: 'static>(
        &self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &'a ComponentState {
        #![allow(unused_variables)]
        if (*self.local_state).type_id() == TypeId::of::<NoState>() {
            panic!("error: 'get_local_state' cannot be called for a component whose root element isn't ComponentOutput")
        }
        self.local_state
            .downcast_ref::<ComponentState>()
            .expect(&format!(
            "internal type error: get_local_state expected {:?} ({}), parent component gave {:?}",
            TypeId::of::<ComponentState>(),
            type_name::<ComponentState>(),
            (*self.local_state).type_id(),
        ))
    }

    // TODO - add methods
    // use_lifecycle
    // get_vdom_context
}

/// Context required by [`VirtualDom::reconcile`]
pub struct ReconcileCtx<'a, 'b, 'c, 'd, 'e> {
    pub event_ctx: &'a mut EventCtx<'d, 'e>,
    pub data: &'b mut DruidAppData,
    pub env: &'c Env,
}

pub struct ProcessEventCtx<'e, 's> {
    pub event_queue: &'e mut dyn Any,
    pub state: &'s mut dyn Any,
}

impl<'e, 's> ProcessEventCtx<'e, 's> {
    pub fn event_queue<ComponentEvent: 'static, ComponentState: 'static>(
        &mut self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &mut Vec<ComponentEvent> {
        #![allow(unused_variables)]
        let type_id = (*self.event_queue).type_id();
        self.event_queue
            .downcast_mut::<Vec<ComponentEvent>>()
            .expect(&format!(
                "internal type error: event handler expected {:?} ({}), parent component gave {:?}",
                TypeId::of::<Vec<ComponentEvent>>(),
                type_name::<Vec<ComponentEvent>>(),
                type_id,
            ))
    }

    pub fn state<ComponentEvent: 'static, ComponentState: 'static>(
        &mut self,
        md: Metadata<ComponentEvent, ComponentState>,
    ) -> &mut ComponentState {
        #![allow(unused_variables)]
        let type_id = (*self.event_queue).type_id();
        self.state.downcast_mut::<ComponentState>().expect(&format!(
            "internal type error: event handler expected {:?} ({}), parent component gave {:?}",
            TypeId::of::<ComponentState>(),
            type_name::<ComponentState>(),
            type_id,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::metadata::NoEvent;
    use test_env_log::test;

    #[derive(Debug, PartialEq, Eq)]
    struct MyEvent(i32);

    #[test]
    fn event_queue() {
        let md: Metadata<MyEvent, NoState> = Default::default();

        let mut event_queue = Vec::<MyEvent>::new();
        let mut ctx = ProcessEventCtx {
            event_queue: &mut event_queue,
            state: &mut NoState,
        };
        ctx.event_queue(md).push(MyEvent(42));

        assert_eq!(event_queue, vec![MyEvent(42)],);
    }

    #[test]
    fn state() {
        let md: Metadata<NoEvent, i64> = Default::default();

        let mut state = 12345_i64;
        let mut ctx = ProcessEventCtx {
            event_queue: &mut Vec::<NoEvent>::new(),
            state: &mut state,
        };

        assert_eq!(*ctx.state(md), 12345_i64,);

        *ctx.state(md) = 123_i64;
        assert_eq!(state, 123_i64,);
    }
}
