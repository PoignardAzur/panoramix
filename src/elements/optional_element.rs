use crate::ctx::{ProcessEventCtx, ReconcileCtx};
use crate::element_tree::{Element, VirtualDom};
use crate::glue::GlobalEventCx;
use crate::metadata::{NoEvent, NoState};

use either::{Either, Left, Right};
use tracing::{debug_span, info, instrument};
use tracing_unwrap::OptionExt;

impl<Child: Element> Element for Option<Child> {
    type Event = NoEvent;
    type ComponentState = NoState;
    type AggregateChildrenState = Option<Child::AggregateChildrenState>;
    type BuildOutput = Option<Child::BuildOutput>;

    #[instrument(name = "Option", skip(self, prev_state))]
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState) {
        if let Some(child) = self {
            let (output, state) = child.build(prev_state.unwrap_or_default());
            (Some(output), Some(state))
        } else {
            (None, None)
        }
    }
}

impl<Child: VirtualDom> VirtualDom for Option<Child> {
    type Event = NoEvent;
    type AggregateChildrenState = Option<Child::AggregateChildrenState>;
    type TargetWidgetSeq = Option<Child::TargetWidgetSeq>;

    #[instrument(name = "Option", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        if let Some(child) = self {
            Some(child.init_tree())
        } else {
            None
        }
    }

    #[instrument(name = "Option", skip(self, other, widget_seq, ctx))]
    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        match (self, other) {
            (Some(child), Some(other_child)) => {
                child.reconcile(other_child, &mut widget_seq.as_mut().unwrap_or_log(), ctx);
            }
            (Some(child), None) => {
                debug_span!("init_tree").in_scope(|| {
                    info!("creating child");
                    *widget_seq = Some(child.init_tree());
                });
            }
            (None, Some(_other_child)) => {
                info!("removing child");
                *widget_seq = None;
            }
            (None, None) => {}
        }
    }

    #[instrument(name = "Option", skip(self, comp_ctx, children_state, widget_seq, cx))]
    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) {
        if let Some(child) = self.as_ref() {
            child.process_event(
                comp_ctx,
                children_state.as_mut().unwrap_or_log(),
                widget_seq.as_mut().unwrap_or_log(),
                cx,
            );
        }
    }
}

// ----

impl<ChildLeft: Element, ChildRight: Element> Element for Either<ChildLeft, ChildRight> {
    type Event = NoEvent;
    type ComponentState = NoState;
    type AggregateChildrenState =
        Option<Either<ChildLeft::AggregateChildrenState, ChildRight::AggregateChildrenState>>;
    type BuildOutput = Either<ChildLeft::BuildOutput, ChildRight::BuildOutput>;

    #[instrument(name = "Either", skip(self, prev_state))]
    fn build(
        self,
        prev_state: Self::AggregateChildrenState,
    ) -> (Self::BuildOutput, Self::AggregateChildrenState) {
        match self {
            Left(child) => {
                let prev_state = prev_state.map_or(None, |ps| ps.left()).unwrap_or_default();
                let (output, state) = child.build(prev_state);
                (Left(output), Some(Left(state)))
            }
            Right(child) => {
                let prev_state = prev_state.map_or(None, |ps| ps.right()).unwrap_or_default();
                let (output, state) = child.build(prev_state);
                (Right(output), Some(Right(state)))
            }
        }
    }
}

impl<ChildLeft: VirtualDom, ChildRight: VirtualDom> VirtualDom for Either<ChildLeft, ChildRight> {
    type Event = NoEvent;
    type AggregateChildrenState =
        Option<Either<ChildLeft::AggregateChildrenState, ChildRight::AggregateChildrenState>>;
    type TargetWidgetSeq = Either<ChildLeft::TargetWidgetSeq, ChildRight::TargetWidgetSeq>;

    #[instrument(name = "Either", skip(self))]
    fn init_tree(&self) -> Self::TargetWidgetSeq {
        match self {
            Left(child) => Left(child.init_tree()),
            Right(child) => Right(child.init_tree()),
        }
    }

    #[instrument(name = "Either", skip(self, other, widget_seq, ctx))]
    fn reconcile(
        &self,
        other: &Self,
        widget_seq: &mut Self::TargetWidgetSeq,
        ctx: &mut ReconcileCtx,
    ) {
        match (self, &other) {
            (Left(child), Left(other)) => {
                // TODO - Add more detailed log
                let widget_seq = &mut widget_seq.as_mut().left().expect_or_log("The previous value of this element was Left. Expected Left widget.");
                child.reconcile(other, widget_seq, ctx);
            }
            (Right(child), Right(other)) => {
                // TODO - Add more detailed log
                let widget_seq = &mut widget_seq.as_mut().right().expect_or_log("The previous value of this element was Right. Expected Right widget.");
                child.reconcile(other, widget_seq, ctx);
            }

            (Left(child), Right(_other)) => {
                debug_span!("init_tree").in_scope(|| {
                    info!("creating child");
                    *widget_seq = Left(child.init_tree());
                });
            }
            (Right(child), Left(_other)) => {
                debug_span!("init_tree").in_scope(|| {
                    info!("creating child");
                    *widget_seq = Right(child.init_tree());
                });
            }
        }
    }

    #[instrument(name = "Either", skip(self, comp_ctx, children_state, widget_seq, cx))]
    fn process_event(
        &self,
        comp_ctx: &mut ProcessEventCtx,
        children_state: &mut Self::AggregateChildrenState,
        widget_seq: &mut Self::TargetWidgetSeq,
        cx: &mut GlobalEventCx,
    ) {
        match self {
            Left(child) => child.process_event(
                comp_ctx,
                &mut children_state
                    .as_mut()
                    .unwrap_or_log()
                    .as_mut()
                    .left()
                    .unwrap_or_log(),
                widget_seq.as_mut().left().unwrap_or_log(),
                cx,
            ),
            Right(child) => child.process_event(
                comp_ctx,
                &mut children_state
                    .as_mut()
                    .unwrap_or_log()
                    .as_mut()
                    .right()
                    .unwrap_or_log(),
                widget_seq.as_mut().right().unwrap_or_log(),
                cx,
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::elements::button::{Button, ButtonData};
    use crate::elements::label::{Label, LabelData};
    use crate::flex::FlexParams;
    use crate::test_harness::Harness;
    use insta::assert_debug_snapshot;
    use test_env_log::test;

    #[test]
    fn new_option() {
        let mut option_label = Some(Label::new("Hello"));
        let (option_label_data, _) = option_label.clone().build(None);

        assert_debug_snapshot!(option_label);
        assert_debug_snapshot!(option_label_data);
        assert_eq!(option_label_data, Some(LabelData::new("Hello")));

        option_label = None;
        let (option_label_data, _) = option_label.clone().build(None);

        assert_debug_snapshot!(option_label);
        assert_debug_snapshot!(option_label_data);
        assert_eq!(option_label_data, None);
    }

    #[test]
    fn new_either() {
        let mut either_elem = Left(Label::new("Hello"));
        let (either_elem_data, _) = either_elem.clone().build(None);

        assert_debug_snapshot!(either_elem);
        assert_debug_snapshot!(either_elem_data);
        assert_eq!(either_elem_data, Left(LabelData::new("Hello")));

        let button_data = ButtonData {
            text: String::from("World"),
            flex: FlexParams {
                flex: 1.0,
                alignment: None,
            },
            ..Default::default()
        };

        either_elem = Right(Button::new("World"));
        let (either_elem_data, _) = either_elem.clone().build(None);

        assert_debug_snapshot!(either_elem);
        assert_debug_snapshot!(either_elem_data);
        assert_eq!(either_elem_data, Right(button_data));
    }

    #[test]
    fn test_option_widget() {
        let some_label = Some(Label::new("Hello"));

        Harness::run_test_window(some_label, |harness| {
            let root_state = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state);

            let new_label = Some(Label::new("World"));
            harness.update_root_element(new_label);

            let root_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state_2);

            harness.update_root_element(None);

            let root_state_3 = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state_3);

            let other_label = Some(Label::new("Goodbye"));
            harness.update_root_element(other_label);

            let root_state_4 = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state_4);
        });
    }

    #[test]
    fn test_either_widget() {
        type LabelButton = Either<Label, Button>;
        let label: LabelButton = Left(Label::new("Hello"));

        Harness::run_test_window(label, |harness| {
            let root_state = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state);

            let new_label: LabelButton = Left(Label::new("World"));
            harness.update_root_element(new_label);

            let root_state_2 = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state_2);

            let button: LabelButton = Right(Button::new("ThisIsAButton"));
            harness.update_root_element(button);

            let root_state_3 = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state_3);

            // TODO - Test reconcile button
            // let new_button: LabelButton = Right(Button::new("AnotherButton"));
            // harness.update_root_element(new_button);
            // let root_state_4 = harness.get_root_debug_state();
            // assert_debug_snapshot!(root_state_4);

            let final_label: LabelButton = Left(Label::new("Goodbye"));
            harness.update_root_element(final_label);

            let root_state_5 = harness.get_root_debug_state();
            assert_debug_snapshot!(root_state_5);
        });
    }
}
