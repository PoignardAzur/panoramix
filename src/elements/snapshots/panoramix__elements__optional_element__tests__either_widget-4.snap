---
source: src/elements/optional_element.rs
expression: root_state_5

---
RootWidget(
    Column(
        Label(
            "Goodbye",
        ),
    ),
)
